Automattic/wp-calypso
# close-on-escape proposition

**Hi All!** 🎅

## Note: 
> The code has changed a bunch since this description was written. The test plan is the same but the approach is no longer to convert to a HoC, rather a component is used within the closable component :)

~~I've been tinkering with some of the mixins lately and looking at converting them to higher order components...~~  
My original approach to refactoring the close-on-esc mixin was to convert it to a HoC, with some feedback from @dmsnell it turned out that this mixin was better suited to being a nested component.

## Motivation

I'm hoping that the dependancy on mixins can be completely removed by replacing their usage with components (either nestable or higher-order). That would allow for the usage of `React.createClass` to be fully replaced with ES6 syntax & `extends component` / `extends pureComponent`.

## Testing

To test these changes I've been doing the following:

- Go to a blog post
- Select 'Featured Image' > 'Set Featured Image' from the side-bar
  <img width="228" alt="featured-image" src="https://cloud.githubusercontent.com/assets/4335450/21477448/20ec2d3a-cb96-11e6-9094-50eb2c8dba78.png">
- Upload or select an image from your media library
- follow through to the image editor for that image
  ![image-editor](https://cloud.githubusercontent.com/assets/4335450/21477547/f941f87c-cb96-11e6-896d-221121b5ce22.gif)
- There's now a 'stack' of escapable components, `dialog-base` and `image-editor`
- Press esc once, the image editor should close.
- Press esc again, the media library (uses dialog-base) should now close too.
- Repeat this process but hold 'esc' down.
- Both components should close in succession.
